lib/active_storage_validations/aspect_ratio_validator.rb

```diff
@@ -1,118 +1,150 @@
-      errors_options[:aspect_ratio] = flat_options[:with]
-      add_error(record, attribute, :aspect_ratio_not_landscape, **errors_options)
-    end
-
-    def validate_regex_aspect_ratio(record, attribute, attachable, flat_options, metadata)
-      flat_options[:with] =~ ASPECT_RATIO_REGEX
-      x = $1.to_i
-      y = $2.to_i
-
-      return if x > 0 && y > 0 && (x.to_f / y).round(PRECISION) == (metadata[:width].to_f / metadata[:height]).round(PRECISION)
-
-      errors_options = initialize_error_options(options, attachable)
-      errors_options[:aspect_ratio] = "#{x}:#{y}"
-      add_error(record, attribute, :aspect_ratio_is_not, **errors_options)
-    end
-
-    def ensure_at_least_one_validator_option
-      unless AVAILABLE_CHECKS.any? { |argument| options.key?(argument) }
-        raise ArgumentError, 'You must pass :with to the validator'
-      end
-    end
-
-    def ensure_aspect_ratio_validity
-      return true if options[:with]&.is_a?(Proc)
-
-      unless NAMED_ASPECT_RATIOS.include?(options[:with]) || options[:with] =~ ASPECT_RATIO_REGEX
-        raise ArgumentError, <<~ERROR_MESSAGE
-          You must pass a valid aspect ratio to the validator
-          It should either be a named aspect ratio (#{NAMED_ASPECT_RATIOS.join(', ')})
-          Or an aspect ratio like 'is_16_9' (matching /#{ASPECT_RATIO_REGEX.source}/)
-        ERROR_MESSAGE
-      end
-    end
-  end
-end
+# frozen_string_literal: true
+
+require_relative 'shared/asv_active_storageable'
+require_relative 'shared/asv_attachable'
+require_relative 'shared/asv_errorable'
+require_relative 'shared/asv_optionable'
+require_relative 'shared/asv_symbolizable'
+
+module ActiveStorageValidations
+  class AspectRatioValidator < ActiveModel::EachValidator # :nodoc
+    include ASVActiveStorageable
+    include ASVAttachable
+    include ASVErrorable
+    include ASVOptionable
+    include ASVSymbolizable
+
+    AVAILABLE_CHECKS = %i[with in].freeze
+    NAMED_ASPECT_RATIOS = %i[square portrait landscape].freeze
+    ASPECT_RATIO_REGEX = /is_([1-9]\d*)_([1-9]\d*)/.freeze
+    ERROR_TYPES = %i[
+      aspect_ratio_not_square
+      aspect_ratio_not_portrait
+      aspect_ratio_not_landscape
+      aspect_ratio_is_not
+      aspect_ratio_invalid
+      image_metadata_missing
+    ].freeze
+    PRECISION = 3.freeze
+
+    def check_validity!
+      ensure_at_least_one_validator_option
+      ensure_aspect_ratio_validity
+    end
+
+    def validate_each(record, attribute, _value)
+      return if no_attachments?(record, attribute)
+
+      validate_changed_files_from_metadata(record, attribute)
+    end
+
+    private
+
+    def is_valid?(record, attribute, attachable, metadata)
+      flat_options = set_flat_options(record)
+
+      return if image_metadata_missing?(record, attribute, attachable, flat_options, metadata)
+
+      aspect_ratios = aspect_ratios(flat_options).compact
+      errors = add_errors(aspect_ratios, metadata)
```

Comment:
Same here, you named for method `add_errors`, but we already have a method doing this (`add_error(record, attribute, error, **errors_options)`).
So, at first, reading your code I am thinking it's doing the same, ie adding the error to the record, but in reality it's not adding any error, it's building possible error message keys.

I was first suggesting naming your method `build_error_keys`. But then, reading the rest of the code, I got the real purpose of your method => validating the fact that the attachable aspect ratio at least passes 1 @allowed_aspect_ratios.

Maybe it could be a better idea to do things like the `content_type_validator` is doing:
```ruby
    def authorized_aspect_ratio?(record, attribute, attachable, metadata)
      attachable_aspect_ratio_is_authorized = @authorized_aspect_ratios.any? do |authorized_aspect_ratio|
        case authorized_aspect_ratio
        when :square then valid_square_aspect_ratio?(metadata)
        ... same for all named aspect ratio
        when ASPECT_RATIO_REGEX then valid_regex_aspect_ratio?(authorized_aspect_ratio, metadata)
        end
      end

      return true if attachable_aspect_ratio_is_authorized

      ... set the error according the current aspect ratio
    end
```

Reply:
and that's it, you do not need the reste of the `is_valid?` method anymore, the logic is now in `authorized_aspect_ratio?`